### DISCUSSION TITLE
generate otp

### DISCUSSION THREAD

**Discussion by krish-adi** [supabase/supabase#30365] [Feature Requests]
auth-js has a function to generate a `magiclink` given an email address which can be used in the server side for custom workflows on authentication. i would like a function that can generate an otp or a token given a phone number (instead of providing an email address here) in the srver side for custom handling of the authentication.

**Comment by GaryAustin1** [comment]
A user here https://discord.com/channels/839993398554656828/1303811204732424192/1303863827523178579 came up with a work around using the auth hooks.   Not pretty but works with phone.

**Reply by dannyhyatt** [reply]
Copying my solution here for those who don't want to join the Discord (and this version uses upsert rather than insert):

```
CREATE TABLE public.otp_codes (
  phone_number text not null unique primary key,
  otp_code text not null
);

CREATE OR REPLACE FUNCTION send_sms(event jsonb) RETURNS jsonb AS $$
BEGIN

  INSERT INTO public.otp_codes (phone_number, otp_code) VALUES (event->'user'->'phone', event->'sms'->'otp')
    ON CONFLICT(phone_number) DO UPDATE SET otp_code = (event->'sms'->'otp');

  RETURN event;

END;
$$ LANGUAGE plpgsql SECURITY DEFINER;
```

Then in your dashboard add this function as an auth hook. Also, I'm using this for load testing so I'm going to remove all this when I'm done. You should probably modify this solution so that it's secure if you're going to be using it in production. Notice also I've used `SECURITY DEFINER` for testing purposes, which the docs recommend not to use.

I would like to see a first-party solution which would make this a lot easier.
